core/src/main/java/io/questdb/network/PipeSuspendEvent.java

```diff
@@ -0,0 +1,72 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2022 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.network;
+
+/**
+ * pipe(2)-based suspend event object. Used on OS X and BSD in combination with kqueue.
+ */
+public class PipeSuspendEvent extends SuspendEvent {
+
+    private final KqueueFacade kqf;
+    private final int readEndFd;
+    private final int writeEndFd;
+
+    public PipeSuspendEvent(KqueueFacade kqf) {
+        this.kqf = kqf;
+        long fds = kqf.pipe();
+        if (fds < 0) {
+            throw NetworkError.instance(kqf.getNetworkFacade().errno(), "cannot create eventfd event");
+        }
+        this.readEndFd = (int) (fds >>> 32);
+        this.writeEndFd = (int) fds;
+        kqf.getNetworkFacade().bumpFdCount(readEndFd);
+        kqf.getNetworkFacade().bumpFdCount(writeEndFd);
+    }
+
+    @Override
+    public void _close() {
+        kqf.getNetworkFacade().close(readEndFd);
+        kqf.getNetworkFacade().close(writeEndFd);
+    }
+
+    @Override
+    public boolean checkTriggered() {
+        return kqf.readPipe(readEndFd) == 1;
+    }
+
+    @Override
+    public int getFd() {
+        return readEndFd;
```

Comment:
what is the purpose of word "end" here?

Reply:
A pipe has two ends: read end is the the one you use to read data from and the write one you use to write it. See the official [documentation](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/pipe.2.html) from Apple (Linux uses the same terminology).

Reply:
thanks 👍